### DISCUSSION TITLE
Unstable features

### DISCUSSION THREAD

**Discussion by Mallets** [eclipse-zenoh/roadmap#65] [Ideas and new features]
I'm creating this discussion departing from the original comment here: https://github.com/eclipse-zenoh/zenoh-c/pull/94#issuecomment-1317133699

As of today, in Zenoh Rust we have the `unstable` feature that allows to introduce experimental capability into Zenoh.
While this approach works fine for Rust, it would be nice to have a similar behaviour in all the bindings:
- https://github.com/eclipse-zenoh/zenoh-c
- https://github.com/eclipse-zenoh/zenoh-python
- https://github.com/eclipse-zenoh/zenoh-cpp

And zenoh-pico:
- https://github.com/eclipse-zenoh/zenoh-pico 

It has to be decided:
- How to support unstable features in a language friendly way. E.g. using `#ifdef` macros or `pragma warnings` in C. What about Python?
- How to adapt Zenoh's CI to support the testing of unstable features across all Zenoh repos
- How to include/exclude unstable features in the Zenoh releases

**Comment by p-avital** [comment]
# C

The feature should be enabled via cmake, as the flags passed to cargo may need to change too. These changes may lead to changes in some structure layouts, so the headers may get tricky. Alternatively, we can chose to always link with `unstable`, which does simplify workflow quite a bit.

# Python

Here we have a few options. Whatever we pick, the python bindings should be built from `unstable`.

A global could be used to let the user opt into `unstable` at runtime. We can use decorators to mark classes/functions as unstable. These decorators can wrap the functions with checks to that opt-in variable and either throw an exception or log a warning if their target is called without having opted into `unstable`.

**Reply by Mallets** [reply]
A reasonable path seems to always build bindings from zenoh `unstable` and then having language-specific ways to enable them in the API:
- C: link to rust `unstable` and use `Cmake` to compile the actual API
- Python: link to rust `unstable` and use decorators. @p-avital do you have an example in mind?
